server/driver/wivrn_hmd.cpp

```diff
@@ -274,6 +292,16 @@ void wivrn_hmd::update_battery(const from_headset::battery & new_battery)
 	battery = new_battery;
 }
 
+void wivrn_hmd::update_fb_face2(const from_headset::fb_face2 & new_face)
+{
+	memcpy(&face.face_expression_set2_fb.weights, new_face.weights.data(), sizeof(float) * new_face.weights.size());
+	memcpy(&face.face_expression_set2_fb.confidences, new_face.confidences.data(), sizeof(float) * new_face.confidences.size());
+	face.face_expression_set2_fb.is_valid = new_face.is_valid;
+	face.face_expression_set2_fb.is_eye_following_blendshapes_valid = new_face.is_eye_following_blendshapes_valid;
+	face.face_expression_set2_fb.data_source = XRT_FACE_TRACKING_DATA_SOURCE2_VISUAL_FB;
+	face.face_expression_set2_fb.sample_time_ns = os_monotonic_get_ns();
+}
+
```

Comment:
This is not thread safe, and considering the comment on the prediction time, I think deriving from `history` could be better.